db_go18.go

```diff
@@ -53,12 +53,16 @@ func (c *conn) beginTxOnce(ctx context.Context, done <-chan struct{}) (*sql.Tx,
 	}
 	go func() {
 		select {
-		case <-ctx.Done():
-			// operation was interrupted by context cancel, so we cancel parent as well
-			c.cancel()
 		case <-done:
 			// operation was successfully finished, so we don't close ctx on tx
 		case <-c.ctx.Done():
+		default:
+			select {
+			case <-ctx.Done():
+				// operation was interrupted by context cancel, so we cancel parent as well
+				c.cancel()
+			default:
```

Comment:
This change doesn't look correct to me.  If I'm reading it correctly, we'll only ever check for `<-ctx.Done()` and `<-c.ctx.Done()` once, upon initialization, and if neither of those return a value, we fall to the `default` case where only `<-ctx.Done()` is ever checked.

What probably makes more sense to me would be something like:

```golang
		select {
		case <-ctx.Done():
			select {
			case <-done():
				// the operation successfully finished at the "same time" as context cancellation, so we won't close ctx on tx
			default:
				// operation was interrupted by context cancel, so we cancel parent as well
				c.cancel()
			}
		case <-done:
			// operation was successfully finished, so we don't close ctx on tx
		case <-c.ctx.Done():
		}
```

Does that make sense?

Reply:
I'm not sure if `<-c.ctx.Done()` should also be checked in the inner select.  I'm not familiar enough with this code to understand the full implications yet.  If someone who's not more familiar can't review this code in a timely fashion, I'll take the time to learn it to make that determination :)

Reply:
Maybe your suggestion is correct.
I will commit the fix after I run the test.